pkg/platform/runtime/setup/setup.go

```diff
@@ -625,3 +618,70 @@ func formatSolverError(serr *apimodel.SolverError) error {
 	}
 	return err
 }
+
+func (s *Setup) completeSetup(artifacts []artifact.ArtifactID) error {
+	edGlobal, err := s.store.UpdateEnviron(artifacts)
+	if err != nil {
+		return errs.Wrap(err, "Could not save combined environment file")
+	}
+
+	// Create executors
+	execPath := ExecDir(s.target.Dir())
+	if err := fileutils.MkdirUnlessExists(execPath); err != nil {
+		return locale.WrapError(err, "err_deploy_execpath", "Could not create exec directory.")
+	}
+
+	exePaths, err := edGlobal.ExecutablePaths()
+	if err != nil {
+		return locale.WrapError(err, "err_deploy_execpaths", "Could not retrieve runtime executable paths")
+	}
+
+	exec := executor.NewWithBinPath(s.target.Dir(), execPath)
+	if err := exec.Update(exePaths); err != nil {
+		return locale.WrapError(err, "err_deploy_executors", "Could not create executors")
+	}
+
+	if err := s.store.MarkInstallationComplete(s.target.CommitUUID()); err != nil {
+		return errs.Wrap(err, "Could not mark install as complete.")
+	}
+
+	return nil
+}
+
+func (s *Setup) installFromDir() error {
+	artifactsDir := s.target.InstallFromDir()
+	if artifactsDir == nil {
+		return errs.New("Cannot install from a directory that is nil")
+	}
+
+	artifacts, err := fileutils.ListDir(*artifactsDir, false)
+	if err != nil {
+		return errs.Wrap(err, "Cannot read from directory to install from")
+	}
+	s.events.TotalArtifacts(len(artifacts))
+	logging.Debug("Found %d artifacts to install from '%s'", len(artifacts), artifactsDir)
+
+	artifactIDs := make([]artifact.ArtifactID, len(artifacts))
+
+	setup := installer.NewSetup(s.store)
+
+	bgErrs, aggregatedErr := aggregateErrors()
+	mainthread.Run(func() {
+		defer close(bgErrs)
+		wp := workerpool.New(MaxConcurrency)
+		for i, a := range artifacts {
+			filename := a.Path()
+			filenameNoExt := filepath.Base(filename[0:strings.Index(filename, ".")])
+			artifactID := artifact.ArtifactID(filenameNoExt)
+			wp.Submit(s.setupArtifactSubmitFunction(artifact.ArtifactDownload{artifactID, filename, "", ""}, setup, bgErrs))
+			artifactIDs[i] = artifactID
+		}
+		wp.StopWait()
+	})
+
+	if err = <-aggregatedErr; err != nil {
+		return err
+	}
+
+	return s.completeSetup(artifactIDs)
+}
```

Comment:
This function is a bit difficult to grok. Please add comments and employ whitespace.